### ISSUE TITLE
Non-interactive "drive init"

### ISSUE THREAD

**Issue by snoopdouglas** [odeke-em/drive#702]
Is there any way that this could be achieved without having to use a web browser and prompts? It'd be very useful for automation.

For example:

```
$ drive init --email="$GOOGLE_EMAIL" --password="$GOOGLE_PASSWORD"

Authenticating... done.
```

Apologies if this has already been suggested; couldn't find it as an issue.

**Comment by odeke-em** [comment]
Hello there @snoopdouglas, thank you for the question and welcome to drive.

In regards to passing in Google email and password, I don't think that's a direction I'd want to see drive go. IMHO, the glaring security holes that involve a user entering their password and email locally to be pasted in a shell are immense.
A password is a master key to all their privacy and every service attached to the owner of the credentials. Moreover one key rules them all; capture this password and you rule every service attached to their account.
On the otherhand, an OAuth2.0 token is the product of an identity access to a service attached to a user's account without needing the direct usage of their master credentials. OAuth2.0 tokens are revokable, can be generated more than once and many can be active simultaneously; they give a limited scope to a very small subset of privileges thereby reducing attack/compromise surfaces greatly. The human intervention required to grant OAuth access also provides an additional layer of security and choice vs automating and blinding accepting. 


#### Exhibits
* For a simple example, even copying and pasting a password to pass it into a shell means it can be viewed in the process list e.g
```shell
$ ps aux | grep email
```
* Their password and username could be viewed in their terminal's history.
* Things copied into paste buffers can be accessed remotely and by other services running on their computer.
* Even with password vaults and managers, password security is still a problem.


In short, privacy is a subject that is very touchy and very hard to handle;  I think it would compromise the security of our users so unfortunately I don't think this is a feature that will be added to drive. Google Security would probably revoke this app totally if we were to try handling passwords :(
Even worse, I don't have any security expertise to be an authority on this topic.

However, if you and anyone else has other suggestions for making automation easy, for sure am all ears.
Thanks.

**Comment by snoopdouglas** [comment]
Hi @odeke-em - that's fine, I understand this - we're not talking about "copying and pasting" in an interactive sense here at all, though.

The environment I'm considering is one where these would be set as environment variables on a VPS. So if someone were to gain access (through eg. stealing an authorised key), the email and password be there in plain sight anyway; they wouldn't have to trawl through `ps aux`.

If this were to be automated, we'd expect such a Google account to simply be a service account for the purpose of having Google Drive space, and nothing more. Perhaps a disclaimer for this could be put together?

If it is possible to interact with drive by simply using an API token and nothing more, I would agree that this would be the better option, though. Do you know whether Google offers such a thing?

**Comment by snoopdouglas** [comment]
Considering better options for handling the username and password, though: how easy would it be to write a script that the user could run *on their workstation* to produce an OAuth key for drive, such that this key could be added to a vault? Can these keys be reused?

**Comment by odeke-em** [comment]
```
If it is possible to interact with drive by simply using an API token and nothing more, I would agree 
that this would be the better option, though. Do you know whether Google offers such a thing?
```
Good question. I think `gcloud` a commandline utility for Google Cloud services has the ability to  manage OAuth tokens with `gcloud auth` on the commandline and that I think could be automated, so we potentially could look into that. So far in drive, what we can automate is something like
```shell
$ drive init --automated
```
which will open up a browser page for you, fire up a server and HTML page that will accept credentials on only that page and then initialize using the credentials it got after the human specified what permissions to grant drive. But still a human would have to review the permissions requested and grant their choice. The only automation here will be no more need to first copy the link on the terminal, manually open a browser window, paste it in there, then copy back the authorization token and paste it back into the terminal. Does this sound reasonable?

```
Considering better options for handling the username and password, though: how easy would it be 
to write a script that the user could run on their workstation to produce an OAuth key for drive, 
such that this key could be added to a vault? Can these keys be reused?
```
Yeah, a friend once suggested to me that I could look into using vaults for storing keys. We were looking at ways of trying to solve https://github.com/odeke-em/drive/issues/44.
In regards to automating this, what I can see off the top of my head would still require human intervention to accept and give permissions for the OAuth token.

Please let me know what you think.

**Comment by snoopdouglas** [comment]
`gcloud` would be a good solution (and as an optional dependency could make this even simpler!)

What about [gcloud-golang](https://github.com/GoogleCloudPlatform/gcloud-golang) - could we use this?
